lesson05/shapes/circle_test.go

```diff
@@ -0,0 +1,151 @@
+		want   error
+	}{
+		{"positive radius", fields{radius: 1}, args{r: 1}, nil},
+		{"zero radius", fields{radius: 0}, args{r: -1}, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", 0.0)},
+		{"negative radius", fields{radius: -1}, args{r: -1}, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", -1.0)},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Circle{
+				radius: tt.fields.radius,
+			}
+			if got := c.SetRadius(tt.args.r); got == nil && tt.want != nil || got != nil && tt.want == nil {
+				t.Errorf("Circle.SetRadius() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCircle_GetRadius(t *testing.T) {
+	type fields struct {
+		radius float64
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		want    float64
+		wantErr error
+	}{
+		{"positive radius", fields{radius: 1}, 1, nil},
+		{"zero radius", fields{radius: 0}, -1, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", 0.0)},
+		{"negative radius", fields{radius: -1}, -1, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", -1.0)},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := Circle{
+				radius: tt.fields.radius,
+			}
+			got, got1 := c.GetRadius()
+			if got != tt.want {
+				t.Errorf("Circle.GetRadius() got = %v, want %v", got, tt.want)
+			}
+			if fmt.Sprint(got1) != fmt.Sprint(tt.wantErr) {
+				t.Errorf("Circle.GetRadius() gotErr = %v, want %v", got1, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestCircle_String(t *testing.T) {
+	type fields struct {
+		radius float64
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		want   string
+	}{
+		{"positive radius", fields{radius: 1}, "\nCircle: radius 1.00"},
+		{"zero radius", fields{radius: 0}, "radius needs to be positive and non-zero, got: 0.000"},
+		{"negative radius", fields{radius: -1}, "radius needs to be positive and non-zero, got: -1.000"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := Circle{
+				radius: tt.fields.radius,
+			}
+			if got := c.String(); got != tt.want {
+				t.Errorf("Circle.String() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCircle_Perimeter(t *testing.T) {
+	type fields struct {
+		radius float64
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		want    float64
+		wantErr error
+	}{
+		{"positive radius", fields{radius: 1}, 2 * math.Pi, nil},
+		{"zero radius", fields{radius: 0}, -1, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", 0.0)},
+		{"negative radius", fields{radius: -1}, -1, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", -1.0)},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := Circle{
+				radius: tt.fields.radius,
+			}
+			got, got1 := c.Perimeter()
+			if got != tt.want {
+				t.Errorf("Circle.Perimeter() got = %v, want %v", got, tt.want)
+			}
+			if fmt.Sprint(got1) != fmt.Sprint(tt.wantErr) {
```

Comment:
not the best way of comparing errors. Errors have a method `Error()`, which returns a string. These strings are better to be used in these cases. Of course you will encounter a case, when error is nil, so you should also make a separate check.
There are similar occurrences throughout other tests here. You should fix them as well

Reply:
fixed all of the statements, now i have seperate checks for when got == nil and when got  != nil